src/net_processing.cpp

```diff
@@ -2402,15 +2402,20 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
 
         std::map<uint256, std::pair<NodeId, std::list<QueuedBlock>::iterator> >::iterator blockInFlightIt = mapBlocksInFlight.find(pindex->GetBlockHash());
         bool fAlreadyInFlight = blockInFlightIt != mapBlocksInFlight.end();
+        bool in_flight_from_same_peer = fAlreadyInFlight ? blockInFlightIt->second.first == pfrom->GetId() : false;
 
         if (pindex->nStatus & BLOCK_HAVE_DATA) // Nothing to do here
             return true;
 
-        if (pindex->nChainWork <= chainActive.Tip()->nChainWork || // We know something better
+        if (pindex->nChainWork <= chainActive.Tip()->nChainWork || // We know same or better
                 pindex->nTx != 0) { // We had this block at some point, but pruned it
-            if (fAlreadyInFlight) {
-                // We requested this block for some reason, but our mempool will probably be useless
+            if ((fAlreadyInFlight && in_flight_from_same_peer)
+                    || pindex->nChainWork == chainActive.Tip()->nChainWork) {
+                // We requested this block for some reason from the same peer
+                // or just want same height block, but our mempool will probably be useless
                 // so we just grab the block via normal getdata
+                std::list<QueuedBlock>::iterator* queuedBlockIt = nullptr;
```

Comment:
This appears to be unused. That said, I think we should consider handling if peer accidentally sends us the block twice, thus queuedBlockIt->partialBlock is set after we call MarkBlockAsInFlight (probably by just wiping the partialBlock).

Reply:
Ok, replicated the logic from a few lines below covering the same case.